src/block.rs

```diff
@@ -170,46 +170,93 @@ impl BlockBuilder {
-        assert!(builder.add_value(b"key3", b"value3", gen_empty_attrs()));
-        let block = builder.build().unwrap();
-        let encoded = block.encode();
-        let _decoded = Block::decode(encoded);
-    }
 
-    #[test]
-    fn test_block_size() {
-        let mut builder = BlockBuilder::new(4096);
-        assert!(builder.add_value(b"key1", b"value1", gen_attrs(1)));
-        assert!(builder.add_value(b"key2", b"value2", gen_attrs(1)));
-        let block = builder.build().unwrap();
-        assert_eq!(73, block.size());
+        for (key, value) in &test_case.entries {
+            let attrs = if test_case.use_timestamped_attrs {
+                gen_attrs(test_case.timestamp.unwrap_or(1)) // Use timestamp value or default to 1
+            } else {
+                gen_empty_attrs()
+            };
 
-        let mut builder = BlockBuilder::new(4096);
-        assert!(builder.add_value(b"key1", b"value1", gen_empty_attrs()));
-        assert!(builder.add_value(b"key2", b"value2", gen_empty_attrs()));
-        let block = builder.build().unwrap();
-        assert_eq!(57, block.size());
+            match value {
+                Some(v) => assert!(builder.add_value(key, v, attrs)),
+                None => assert!(builder.add_tombstone(key, gen_empty_attrs())),
+            }
+        }
+        builder.build().expect("Failed to build block")
+    }
+
+    #[rstest]
+    #[case(BlockTestCase {
+        name: "test_block",
+        entries: vec![
+            (b"key1", Some(b"value1")),
+            (b"key1", Some(b"value1")),
+            (b"key2", Some(b"value2")),
+        ],
+        expected_size: None,
+        use_timestamped_attrs: false,
+        timestamp: None,
+    })]
+    #[case(BlockTestCase {
+        name: "block_with_tombstone",
+        entries: vec![
+            (b"key1", Some(b"value1")),
+            (b"key2", None),
+            (b"key3", Some(b"value3")),
+        ],
+        expected_size: None,
+        use_timestamped_attrs: false,
+        timestamp: None,
+    })]
+    #[case(BlockTestCase {
+        name: "block_size_with_attrs",
+        entries: vec![
+            (b"key1", Some(b"value1")),
+            (b"key2", Some(b"value2")),
+        ],
+        expected_size: Some(73),
+        use_timestamped_attrs: true,
+        timestamp: Some(1),
+    })]
+    #[case(BlockTestCase {
+        name: "block_size_with_empty_attrs",
+        entries: vec![
+            (b"key1", Some(b"value1")),
+            (b"key2", Some(b"value2")),
+        ],
+        expected_size: Some(57),
+        use_timestamped_attrs: false,
+        timestamp: None,
+    })]
+    fn test_block(#[case] test_case: BlockTestCase) {
+        let block = build_block(&test_case);
+        let encoded = block.encode();
+        let decoded = Block::decode(encoded);
+        assert_eq!(block.data, decoded.data);
+        assert_eq!(block.offsets, decoded.offsets);
+        if let Some(expected_size) = test_case.expected_size {
```

Comment:
Sorry to add this one on, but after re-checking, I think we should just require size for all test cases (i.e. not make it Option). Mind tweaking the test to do this, and filling in the missing sizes?

Reply:
Sounds good , if we consider the size to be checked for all test cases that doesn't have it's snapshot file will be re-created as well i.e. we will have a new snapshot moving forward.

Reply:
Yep, sounds good.